### ISSUE TITLE
Allow client-authentication to limit who downloads the software

### ISSUE THREAD

**Issue by dragetd** [setlog/trivrost#68]
**Is your feature request related to a problem? Please describe.**
We received external input, that one use-case would be to limit who is able to access the URLs and not let everyone download the software.

**Describe the solution you'd like**
Discuss some possibilities to limit the client access.

**Describe alternatives you've considered**
One of our benefits (which we should document!) is, that our update-server only requires to be a "simple" HTTP server. This means we should not rely on custom server code but stick to HTTP semantics that can be configured with common webservers.

Alternative 1: Client Basic Auth
Relatively easy to use in go (https://golang.org/pkg/net/http/#Request.BasicAuth). The client would need to provide a username and password.
This could be as simple as a default-password that is baked into the client (launcher-config) and prevents access from third-parties who never got the launcher. Of course, once the password is 'leaked', it provides no more protection. This is not a very secure scheme for anything, but also simple to implement.

Per-client passwords would require some kind of backend or infrastructure to manage these. Sounds a bit out of scope for trivrost, but I am open for ideas.

ENFORCES HTTPS.

Alternative 2: Client certificates
Client certificates are a kind of individual client credentials, but allow generating these and providing trust though a CA. It they are signed by the CA, they are trusted and no more backend-logic is needed. Still ,each client would need its own cert. Not sure how this would be done.

Implementing it in go is a little more tedious (= error prone) (https://gist.github.com/michaljemala/d6f4e01c4834bf47a9c4), but possible.

Alternative 3: Identify clients but local client data
Some sort of hash from MAC, serial numbers and whatnot… no idea how exactly it would look like, but I already know this is simple security by obscurity and even more stupid that alternative 1. But again, better than nothing.

Open for more ideas.

**Comment by MMulthaupt** [comment]
Lots of opportunity to reinvent the wheel here. Best we could do as far as I see is react to the [`WWW-Authenticate`](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/WWW-Authenticate) response header and act accordingly. As far as I known, basic auth over TLS is secure. If a single user leaks credentials or the software itself, it's game over immediately, regardless of what authentication method is used. All the onion-design in the world won't change that.

**Comment by dragetd** [comment]
Yeah, but there is still a difference between shared secrets and user certificates. A shared secret is worthless once leaked. A single compromised client certificate does not invalidate the others.

Tho basic-auth will certainly be a good-engouh option for many cases I guess.
